src/core/batch.rkt

```diff
@@ -82,71 +82,173 @@
+      [(list 'sinh args)
+       (define exp-idx (append-node `(exp ,(hash-ref mappings args))))
+       (define one-idx (append-node 1))
+       (define inv-exp-idx (append-node `(/ ,one-idx ,exp-idx)))
+       (define neg-idx (append-node `(neg ,inv-exp-idx)))
+       (define add-idx (append-node `(+ ,exp-idx ,neg-idx)))
+       (define half-idx (append-node 1/2))
+       (hash-set! mappings n (append-node `(* ,half-idx ,add-idx)))]
+      [(list 'tanh args)
+       (define exp-idx (append-node `(exp ,(hash-ref mappings args))))
+       (define one-idx (append-node 1))
+       (define inv-exp-idx (append-node `(/ ,one-idx ,exp-idx)))
+       (define neg-idx (append-node `(neg ,inv-exp-idx)))
+       (define add-idx (append-node `(+ ,exp-idx ,inv-exp-idx)))
+       (define sub-idx (append-node `(+ ,exp-idx ,neg-idx)))
+       (hash-set! mappings n (append-node `(/ ,sub-idx ,add-idx)))]
+      [(list 'asinh args)
+       (define mult-idx (append-node `(* ,(hash-ref mappings args) ,(hash-ref mappings args))))
+       (define one-idx (append-node 1))
+       (define add-idx (append-node `(+ ,mult-idx ,one-idx)))
+       (define sqrt-idx (append-node `(sqrt ,add-idx)))
+       (define add2-idx (append-node `(+ ,(hash-ref mappings args) ,sqrt-idx)))
+       (hash-set! mappings n (append-node `(log ,add2-idx)))]
+      [(list 'acosh args)
+       (define mult-idx (append-node `(* ,(hash-ref mappings args) ,(hash-ref mappings args))))
+       (define -one-idx (append-node -1))
+       (define add-idx (append-node `(+ ,mult-idx ,-one-idx)))
+       (define sqrt-idx (append-node `(sqrt ,add-idx)))
+       (define add2-idx (append-node `(+ ,(hash-ref mappings args) ,sqrt-idx)))
+       (hash-set! mappings n (append-node `(log ,add2-idx)))]
+      [(list 'atanh args)
+       (define neg-idx (append-node `(neg ,(hash-ref mappings args))))
+       (define one-idx (append-node 1))
+       (define add-idx (append-node `(+ ,one-idx ,(hash-ref mappings args))))
+       (define sub-idx (append-node `(+ ,one-idx ,neg-idx)))
+       (define div-idx (append-node `(/ ,add-idx ,sub-idx)))
+       (define log-idx (append-node `(log ,div-idx)))
+       (define half-idx (append-node 1/2))
+       (hash-set! mappings n (append-node `(* ,half-idx ,log-idx)))]
+      [(list op args ...)
+       (hash-set! mappings n (append-node (cons op (map (curry hash-ref mappings) args))))]
+      [(approx spec impl) (hash-set! mappings n (append-node (approx spec (hash-ref mappings impl))))]
+      [_ (hash-set! mappings n (append-node node))]))
+
+  (define roots* (vector-map (curry hash-ref mappings) roots))
+  (define nodes* (list->vector (reverse icache)))
+
+  ; This may be too expensive to handle simple 1/2, 1/3 and 2/3 zombie nodes..
+  #;(remove-zombie-nodes (batch nodes* roots* vars (vector-length nodes*)))
+
+  (batch nodes* roots* vars (vector-length nodes*)))
+
+; The function removes any zombie nodes from batch
+(define (remove-zombie-nodes input-batch)
```

Comment:
This function is not used currently because it is too expensive.
However, it is tested and works correctly